src/QsCompiler/CompilerOptimizations/PureCircuitAPI.fs

```diff
@@ -4,197 +4,175 @@
 module Microsoft.Quantum.QsCompiler.CompilerOptimization.PureCircuitAPI
 
 open System.Collections.Immutable
-open System.Numerics
 open Microsoft.Quantum.QsCompiler.DataTypes
 open Microsoft.Quantum.QsCompiler.SyntaxTokens
 open Microsoft.Quantum.QsCompiler.SyntaxTree
 
 open ComputationExpressions
-open Types
 open Utils
 open Printer
 
 
-/// An identifier for a qubit or for an array of qubits
-type QubitReference = Qubit of int | QubitArray of int
-
 /// Any constant expression
-type Const =
-| UnitValue
-| ValueTuple            of Const list
+type Literal =
 | IntLiteral            of int64
-| BigIntLiteral         of BigInteger
 | DoubleLiteral         of double
-| BoolLiteral           of bool
-| StringLiteral         of string
-| ResultLiteral         of QsResult
 | PauliLiteral          of QsPauli
-| RangeLiteral          of Const * Const
-| ValueArray            of Const list
-| ArrayItem             of Const * Const
-| QubitReference        of QubitReference
+| PauliArray            of QsPauli list
+
+/// Any expression
+type Expression =
+| Literal               of Literal
+| Tuple                 of Expression list
+| Qubit                 of int
```

Comment:
I know I am being nitpicky, but having spent days on hunting down bugs due to similar things: Are you checking that the indices fit into a 32-bit integer?

Reply:
I think I saw that the statement removing should take care of that. Is it also handled if that step is not run?

Reply:
If the index didn't fit into a 32-bit integer, that would mean that the code has at least 2^31 distinct qubit allocations or unknown values. I have to imagine that if the code really did have more than 2 billion distinct expressions, many other parts of the compiler would break first.

Reply:
Probably a fair point. :)

Reply:
Actually, please handle this properly. I did ping the libraries team, because it seemed like a possibility that we might be working with that many qubit allocations on the toffoli simulator, and we are. Thanks!

Reply:
It's not even possible to create lists that long, and a lot of the library functions I'm using only work with integers. Handling this case would be somewhat difficult, and given that this is an extreme edge case that we won't run into for years, I'm not sure it's worth it. What if the compiler throws an error if there would be an overflow?

Reply:
Fixed in latest commit